### ISSUE TITLE
AMP callRemoteString ignores requiresAnswer

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3999]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#3999|
|Type|defect|
|Created|2009-09-04 12:09:54Z|

`BoxDispatcher.callRemoteString` has a `requiresAnswer` argument (and documentation), but it is not passed to `return self._sendBoxCommand(command, box)`. There are also no tests for the case where `requiresAnswer=False`.

Attachments:

* [amp_callRemoteString_requiresAnwerFalse.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6d4/6d438f23010203dfe40bfec8034252e1450aa29f/429347f5efb68cea7bc610805b6060921b93f2fa.patch) (1231 bytes) - added by jesstess on 2009-09-17 04:09:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3999 3999
type__defect defect
reporter__ivank ivank
priority__normal normal
milestone__ 
branch__ 
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1252066194000000 1252066194000000
changetime__1253161863000000 1253161863000000
version__None None
owner__jesstess jesstess
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Fixing description markup.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess removed owner|
|-|-|

The patch adds requiresAnswer to the call to _sendBoxCommand and adds a test case in twisted/test/test_amp.py that callRemoteString returns None if requiresAnswer is False.

Something else came up while looking into this ticket:

It looks like callRemoteString creates a new AmpBox that it passes to _sendBoxCommand, and when you make requiresAnswer False, in _sendBoxCommand only a _command tag gets added to this new box before calling box._sendTo(self.boxSender).

However, the AmpBox._sendTo docstring states that exactly one of _ask, _answer, and _error is a required key. 

Is the docstring incorrect, or is the path that returnFalse forces bad? Other tests in 
test_amp.py, including test_requiresNoAnswer and test_requiresNoAnswerFail, produce AmpBoxes without one of the required keys (ostensibly _ask should be set, as _command is required if _ask is present).

print statements from AmpBox._sendBox while running the tests:
    test_callRemoteStringRequiresAnswerFalse ... AmpBox({'_command': 'WTF'})
                          [OK]
    test_requiresNoAnswer ... AmpBox({'hello': 'world', '_command': 'hello'})
                                             [OK]
    test_requiresNoAnswerFail ... AmpBox({'hello': 'fuck you', '_command': 'hello'})
                                         [OK]

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @jesstess|
|-|-|

Nice patch! Just one comment: please use assertIdentical to compare to None.

Please apply then.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess set status to `closed`|
|-|-|

(In [27346]) Apply amp_callRemoteString_requiresAnwerFalse.patch

Author: jesstess Reviewer: therve Fixes: [#3999](https://github.com/twisted/twisted/issues/3999)

Have BoxDispatcher.callRemoteString pass its requiresAnswer argument to its call to _sendBoxCommand. Also add a unit test for requiresAnswer = False.
